# Starting H2O server

In [1]:
import h2o
from h2o.automl import H2OAutoML
h2o.init()

Checking whether there is an H2O instance running at http://localhost:54321..... not found.
Attempting to start a local H2O server...
; Java HotSpot(TM) 64-Bit Server VM (build 25.171-b11, mixed mode)
  Starting server from C:\Users\sayal\AppData\Roaming\Python\Python36\site-packages\h2o\backend\bin\h2o.jar
  Ice root: C:\Users\sayal\AppData\Local\Temp\tmpmx0izerf
  JVM stdout: C:\Users\sayal\AppData\Local\Temp\tmpmx0izerf\h2o_sayal_started_from_python.out
  JVM stderr: C:\Users\sayal\AppData\Local\Temp\tmpmx0izerf\h2o_sayal_started_from_python.err
  Server is running at http://127.0.0.1:54321
Connecting to H2O server at http://127.0.0.1:54321... successful.


--------------------------  -------------------------------
H2O cluster uptime:         03 secs
H2O cluster timezone:       America/New_York
H2O data parsing timezone:  UTC
H2O cluster version:        3.22.0.2
H2O cluster version age:    3 months and 3 days
H2O cluster name:           H2O_from_python_sayal_9jomju
H2O cluster total nodes:    1
H2O cluster free memory:    3.535 Gb
H2O cluster total cores:    8
H2O cluster allowed cores:  8
H2O cluster status:         accepting new members, healthy
H2O connection url:         http://127.0.0.1:54321
H2O connection proxy:
H2O internal security:      False
H2O API Extensions:         Algos, AutoML, Core V3, Core V4
Python version:             3.6.5 final
--------------------------  -------------------------------

h2o.init with python seems very sensitive the the H2O version. If the H2O cluster version is 3.20.0.1 and the python h2o library is 3.19.0 it will fail so we set strict_version_check=False

If the H2O cluster isn't found h2o.init will start one.

Note that the current script starts each H2O instance on a different port. It's not clear why but should we do this we should choose from only the higher ports.

A port number is a 16-bit unsigned integer, thus ranging from 0 to 65535. There is no reason to choose a port less than 10000.

In [2]:
df = h2o.import_file('./train.csv')

Parse progress: |█████████████████████████████████████████████████████████| 100%


In [16]:
df.describe()

Rows:137
Cols:43




Id Open Date City City Group Type P1 P2 P3 P4 P5 P6 P7 P8 P9 P10 P11 P12 P13 P14 P15 P16 P17 P18 P19 P20 P21 P22 P23 P24 P25 P26 P27 P28 P29 P30 P31 P32 P33 P34 P35 P36 P37 revenue 
 
 
 type int time enum enum enum int real real real int int int int int int int int real int int int int int int int int int int int int real real real real int int int int int int int int int 
 mins 0.0 831513600000.0 1.0 1.0 0.0 3.0 1.0 1.0 1.0 1.0 4.0 4.0 1.0 2.0 3.0 0.0 0.0 0.0 0.0 0.0 1.0 1.0 1.0 1.0 1.0 0.0 0.0 0.0 0.0 1.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1149870.0 
 mean 68.0 1237657296350.3645 4.014598540145983 4.40875912408759 4.3175182481751815 4.3722627737226265 2.007299270072992 3.357664233576642 5.423357664233579 5.153284671532846 5.445255474452556 5.4890510948905105 3.2627737226277373 5.299270072992701 5.0802919708029215 1.416058394160584 1.3868613138686123 1.9416058394160591 1.0364963503649631 1.9416058394160591 4.9051094890510925 4.547445255474453 2.2700729927007313 2.226277372262773 3.4233576642335772 1.3722627737226274 1.2116788321167888 1.4708029197080286 1.1459854014598536 3.222627737226278 3.1350364963503643 2.7299270072992696 1.9416058394160591 2.525547445255475 1.1386861313868615 2.489051094890511 2.0291970802919708 2.2116788321167897 1.1167883211678835 4453532.613138688 
 maxs 136.0 1390608000000.0 12.0 7.5 7.5 7.5 8.0 10.0 10.0 10.0 10.0 10.0 10.0 10.0 7.5 15.0 10.0 15.0 15.0 12.0 25.0 15.0 15.0 5.0 25.0 10.0 10.0 12.5 12.5 12.5 7.5 25.0 15.0 25.0 6.0 24.0 15.0 20.0 8.0 19696939.0 
 sigma 39.69256857397868 127116648579.15549 2.9103906182641897 1.5149002322301595 1.0323372831631499 1.016462285331371 1.2096204458482704 2.134235040936531 2.2968092565212213 1.8585672492102188 1.8347932566223957 1.8475611441015003 1.9107669887212244 1.941667580380517 1.0365269651197466 2.729582859119749 2.398676642450269 3.5058069964611924 2.0306792514853385 3.3005490259814088 5.604466721336747 3.7080413470149143 2.052629768180595 1.2306902697860198 4.55960948362845 2.3041117315830677 2.1331785760383997 2.6120242483769287 2.067038752821938 2.3088064366403795 1.6808869875261374 5.536647490657986 3.51209344173743 5.230116819034456 1.698539518630745 5.165093228930779 3.4362719321608006 4.168210677141203 1.7907675540962253 2576071.595708188 
 zeros 1 0 0 0 1 0 0 0 0 0 0 0 0 0 0 88 88 88 88 88 0 0 0 0 0 88 88 88 89 0 2 88 88 88 88 88 88 88 88 0 
 missing 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 
 0 0.0 1999-07-17 00:00:00 Ä°stanbul Big Cities IL 4.0 5.0 4.0 4.0 2.0 2.0 5.0 4.0 5.0 5.0 3.0 5.0 5.0 1.0 2.0 2.0 2.0 4.0 5.0 4.0 1.0 3.0 3.0 1.0 1.0 1.0 4.0 2.0 3.0 5.0 3.0 4.0 5.0 5.0 4.0 3.0 4.0 5653753.0 
 1 1.0 2008-02-14 00:00:00 Ankara Big Cities FC 4.0 5.0 4.0 4.0 1.0 2.0 5.0 5.0 5.0 5.0 1.0 5.0 5.0 0.0 0.0 0.0 0.0 0.0 3.0 2.0 1.0 3.0 2.0 0.0 0.0 0.0 0.0 3.0 3.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 6923131.0 
 2 2.0 2013-03-09 00:00:00 Diyarbakır Other IL 2.0 4.0 2.0 5.0 2.0 3.0 5.0 5.0 5.0 5.0 2.0 5.0 5.0 0.0 0.0 0.0 0.0 0.0 1.0 1.0 1.0 1.0 1.0 0.0 0.0 0.0 0.0 1.0 3.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 2055379.0 
 3 3.0 2012-02-02 00:00:00 Tokat Other IL 6.0 4.5 6.0 6.0 4.0 4.0 10.0 8.0 10.0 10.0 8.0 10.0 7.5 6.0 4.0 9.0 3.0 12.0 20.0 12.0 6.0 1.0 10.0 2.0 2.0 2.5 2.5 2.5 7.5 25.0 12.0 10.0 6.0 18.0 12.0 12.0 6.0 2675511.0 
 4 4.0 2009-05-09 00:00:00 Gaziantep Other IL 3.0 4.0 3.0 4.0 2.0 2.0 5.0 5.0 5.0 5.0 2.0 5.0 5.0 2.0 1.0 2.0 1.0 4.0 2.0 2.0 1.0 2.0 1.0 2.0 3.0 3.0 5.0 1.0 3.0 5.0 1.0 3.0 2.0 3.0 4.0 3.0 3.0 4316715.0 
 5 5.0 2010-02-12 00:00:00 Ankara Big Cities FC 6.0 6.0 4.5 7.5 8.0 10.0 10.0 8.0 8.0 8.0 10.0 8.0 6.0 0.0 0.0 0.0 0.0 0.0 5.0 6.0 3.0 1.0 5.0 0.0 0.0 0.0 0.0 7.5 5.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 5017319.0 
 6 6.0 2010-10-11 00:00:00 Ä°stanbul Big Cities IL 2.0 3.0 4.0 4.0 1.0 5.0 5.0 5.0 5.0 5.0 2.0 5.0 5.0 3.0 4.0 4.0 3.0 4.0 2.0 4.0 1.0 2.0 1.0 5.0 4.0 4.0 5.0 1.0 3.0 4.0 5.0 2.0 2.0 3.0 5.0 4.0 4.0 5166635.0 
 7 7.0 2011-06-21 00:00:00 Ä°stanbul Big Cities IL 4.0 5.0 4.0 5.0 2.0 3

In [17]:
y = 'revenue'

In [18]:
aml = H2OAutoML(max_runtime_secs = 6000, seed = 1)

In [19]:
splits = df.split_frame(ratios = [0.8], seed = 1)
train = splits[0]
test = splits[1]

In [20]:
aml.train(y = y, training_frame = train, leaderboard_frame =test )

AutoML progress: |████████████████████████████████████████████████████████| 100%


# Leaderboard

Next, we will view the AutoML Leaderboard. Since we did not specify a leaderboard_frame in the H2OAutoML.train() method for scoring and ranking the models, the AutoML leaderboard uses cross-validation metrics to rank the models.

A default performance metric for each machine learning task (binary classification, multiclass classification, regression) is specified internally and the leaderboard will be sorted by that metric. In the case of binary classification, the default ranking metric is Area Under the ROC Curve (AUC). In the future, the user will be able to specify any of the H2O metrics so that different metrics can be used to generate rankings on the leaderboard.

The leader model is stored at aml.leader and the leaderboard is stored at aml.leaderboard.

In [21]:
aml.leaderboard

model_id mean_residual_deviance rmse mse mae rmsle 
 
 
 DeepLearning_grid_1_AutoML_20190225_201636_model_14 1.77414e+12 1.33197e+06 1.77414e+12 1.07649e+06 0.396311 
 DeepLearning_grid_1_AutoML_20190225_201636_model_23 1.87166e+12 1.36809e+06 1.87166e+12 1.06591e+06 0.398778 
 DeepLearning_grid_1_AutoML_20190225_201636_model_31 2.29881e+12 1.51618e+06 2.29881e+12 1.24849e+06 0.392972 
 DeepLearning_grid_1_AutoML_20190225_201636_model_22 2.44535e+12 1.56376e+06 2.44535e+12 1.24131e+06 0.425939 
 DeepLearning_grid_1_AutoML_20190225_201636_model_30 2.65563e+12 1.62961e+06 2.65563e+12 1.3693e+06 0.455223 
 GBM_grid_1_AutoML_20190225_201636_model_1 2.88285e+12 1.6979e+06 2.88285e+12 1.52865e+06 0.49207 
 GBM_grid_1_AutoML_20190225_201636_model_2 2.89848e+12 1.70249e+06 2.89848e+12 1.53041e+06 0.496404 
 DeepLearning_grid_1_AutoML_20190225_201636_model_5 2.99737e+12 1.73129e+06 2.99737e+12 1.46429e+06 0.448689 
 DeepLearning_grid_1_AutoML_20190225_201636_model_25 3.0327e+12 1.74146e+06 3.0327e+12 1.5192e+06 0.493735 
 DeepLearning_grid_1_AutoML_20190225_201636_model_3 3.05941e+12 1.74912e+06 3.05941e+12 1.42313e+06 0.456737

In [22]:
perf = aml.leader.model_performance(test)
perf


ModelMetricsRegression: deeplearning
** Reported on test data. **

MSE: 1774138315871.0547
RMSE: 1331967.8358995966
MAE: 1076485.1875787657
RMSLE: 0.3963110090679296
Mean Residual Deviance: 1774138315871.0547


# Conclusion

Our Evaluation is based on RMSE values
We are getting a RMSE value of 1331967.8358995966 